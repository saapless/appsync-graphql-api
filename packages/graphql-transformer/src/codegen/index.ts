export * from "./ast";
export * from "./printer";
export * from "./document";
