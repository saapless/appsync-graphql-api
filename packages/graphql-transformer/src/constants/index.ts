export const ConnectionDirective = Object.freeze({
  HAS_ONE: "hasOne",
  HAS_MANY: "hasMany",
});

export const ModelDirective = Object.freeze({
  MODEL: "model",
});

export const UtilityDirective = Object.freeze({
  SERVER_ONLY: "serverOnly",
  CLIENT_ONLY: "clientOnly",
  READ_ONLY: "readOnly",
  WRITE_ONLY: "writeOnly",
  FILTER_ONLY: "filterOnly",
  INTERNAL: "internal",
});

export const ScalarType = Object.freeze({
  STRING: "String",
  INT: "Int",
  FLOAT: "Float",
  BOOLEAN: "Boolean",
  ID: "ID",
  LONG: "Long",
});

export const RelationType = Object.freeze({
  ONE_TO_ONE: "oneToOne",
  ONE_TO_MANY: "oneToMany",
  MANY_TO_MANY: "manyToMany",
});
