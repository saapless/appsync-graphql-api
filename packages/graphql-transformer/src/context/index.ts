export * from "./TransformerContext";
export type {
  DescriptorOperation,
  Key,
  KeyOperator,
  KeyValue,
  OperationActionType,
  ResolverDescriptor,
  ReturnType,
} from "./ResolverLoader";
