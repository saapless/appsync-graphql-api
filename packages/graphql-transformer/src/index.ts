export * from "./transformer";
export { definitionFromSource } from "./utils";
