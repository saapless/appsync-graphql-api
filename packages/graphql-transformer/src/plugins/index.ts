export { TransformerPluginBase, type IPluginFactory } from "./TransformerPluginBase";
export { AWSTypesPlugin } from "./AWSTypesPlugin";
export { AuthPlugin } from "./AuthPlugin";
export { ModelPlugin } from "./ModelPlugin";
export { NodeInterfacePlugin } from "./NodeInterfacePlugin";
export { ConnectionPlugin } from "./ConnectionPlugin";
