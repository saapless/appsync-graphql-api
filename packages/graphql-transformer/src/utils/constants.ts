/**
 * DIRECTIVES
 */
export const AUTH_DIRECTIVE = "auth";

export const AWS_AUTH_DIRECTIVE = "aws_auth";
export const AWS_API_KEY_DIRECTIVE = "aws_api_key";
export const AWS_IAM_DIRECTIVE = "aws_iam";
export const AWS_OIDC_DIRECTIVE = "aws_oidc";
export const AWS_LAMBDA_DIRECTIVE = "aws_lambda";
export const AWS_COGNITO_USER_POOLS_DIRECTIVE = "aws_cognito_user_pools";
export const AWS_SUBSCRIBE_DIRECTIVE = "aws_subscribe";

export const HAS_ONE_DIRECTIVE = "hasOne";
export const HAS_MANY_DIRECTIVE = "hasMany";

export const DATA_SOURCE_DIRECTIVE = "dataSource";
export const RESOLVER_DIRECTIVE = "resolver";

export const MODEL_DIRECTIVE = "model";

export const SERVER_ONLY_DIRECTIVE = "serverOnly";
export const CLIENT_ONLY_DIRECTIVE = "clientOnly";
export const READ_ONLY_DIRECTIVE = "readOnly";
export const WRITE_ONLY_DIRECTIVE = "writeOnly";
export const FILTER_ONLY_DIRECTIVE = "filterOnly";

export const DEFAULT_ID_FIELD = "id";
export const DEFAULT_SK_FIELD = "_sk";

export const DEFAULT_SOURCE_INDEX = "bySourceId";
export const DEFAULT_SOURCE_INDEX_FIELD = "sourceId";
export const DEFAULT_TARGET_INDEX = "byTargetId";
export const DEFAULT_TARGET_INDEX_FIELD = "targetId";
export const DEFAULT_TYPENAME_INDEX = "byTypename";
export const DEFAULT_TYPENAME_INDEX_FIELD = "__typename";
