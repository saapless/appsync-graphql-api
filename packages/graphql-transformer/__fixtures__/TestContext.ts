import { TransformerContext, TransformerContextConfig } from "../src/context";

interface TestingTransformerConfig extends TransformerContextConfig {
  allowPrint?: boolean;
}

export class TestContext extends TransformerContext {
  public allowPrint: boolean;
  public files: Map<string, string> = new Map();

  constructor(config: TestingTransformerConfig) {
    super(config);
    this.allowPrint = config.allowPrint ?? false;
    this.files = new Map();
  }

  public override createOutputDirectory(path: string) {
    return path;
  }

  public printFile(filePath: string, content: string) {
    if (this.allowPrint) {
      return super.printFile(filePath, content);
    }

    this.files.set(filePath, content);
  }

  public printScript(filePath: string, content: string) {
    if (this.allowPrint) {
      return super.printScript(filePath, content);
    }

    this.files.set(filePath, content);
  }
}
