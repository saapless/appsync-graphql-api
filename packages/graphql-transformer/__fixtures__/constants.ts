import { DataSourceManagerConfig } from "../src/context/DataSourceManager";

export const TEST_DS_CONFIG = {
  primaryDataSourceName: "TestDataSource",
  dataSources: {
    TestDataSource: {
      type: "DYNAMO_DB",
      config: {
        tableName: "TestTable",
      },
    },
    NoneDataSource: {
      type: "NONE",
    },
  },
} satisfies DataSourceManagerConfig;
