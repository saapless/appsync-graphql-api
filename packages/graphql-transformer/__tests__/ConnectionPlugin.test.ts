import { TEST_DS_CONFIG } from "../__fixtures__/constants";
import { TransformerContext } from "../src/context";
import {
  DirectiveDefinitionNode,
  DocumentNode,
  EnumNode,
  FieldNode,
  InputObjectNode,
  ObjectNode,
} from "../src/definition";
import { ConnectionPlugin } from "../src/plugins";

const schema = /* GraphQL */ `
  type User {
    id: ID!
    username: String!
  }

  type Todo {
    id: ID
    content: String
    resources: Resource @hasMany
  }

  type Document {
    url: String!
  }

  type Message {
    content: String!
  }

  union Resource = Document | Message

  type Query {
    me: User @hasOne(key: { ref: "identity.sub" })
    todos: Todo
      @hasMany(
        key: { ref: "identity.sub" }
        sortKey: { beginsWith: { eq: "Todo" } }
        index: "bySourceId"
      )
  }
`;

describe("ConnectionPlugin", () => {
  const context = new TransformerContext({
    document: DocumentNode.fromSource(schema),
    dataSourceConfig: TEST_DS_CONFIG,
  });

  const plugin = ConnectionPlugin.create(context);

  describe("on run `before` hook", () => {
    beforeAll(() => plugin.before());

    it("adds connection directive definitions", () => {
      expect(context.document.getNode("ConnectionRelationType")).toBeInstanceOf(EnumNode);
      expect(context.document.getNode("SortKeyInput")).toBeInstanceOf(InputObjectNode);
      expect(context.document.getNode("hasOne")).toBeInstanceOf(DirectiveDefinitionNode);
      expect(context.document.getNode("hasMany")).toBeInstanceOf(DirectiveDefinitionNode);
    });

    it("adds scalars filter types", () => {
      expect(context.document.getNode("StringFilterInput")).toBeInstanceOf(InputObjectNode);
      expect(context.document.getNode("IntFilterInput")).toBeInstanceOf(InputObjectNode);
      expect(context.document.getNode("FloatFilterInput")).toBeInstanceOf(InputObjectNode);
      expect(context.document.getNode("BooleanFilterInput")).toBeInstanceOf(InputObjectNode);
      expect(context.document.getNode("IDFilterInput")).toBeInstanceOf(InputObjectNode);
      expect(context.document.getNode("SizeFilterInput")).toBeInstanceOf(InputObjectNode);
      expect(context.document.getNode("ListFilterInput")).toBeInstanceOf(InputObjectNode);
      expect(context.document.getNode("SortDirection")).toBeInstanceOf(EnumNode);
    });
  });

  describe("on normalize node", () => {
    beforeAll(() => {
      plugin.normalize(context.document.getQueryNode());
      plugin.normalize(context.document.getNode("Todo") as ObjectNode);
    });

    it("adds connection keys to nodes", () => {
      expect((context.document.getNode("Todo") as ObjectNode).getField("sourceId")).toBeInstanceOf(
        FieldNode
      );
      expect(
        (context.document.getNode("Document") as ObjectNode).getField("sourceId")
      ).toBeInstanceOf(FieldNode);
      expect(
        (context.document.getNode("Message") as ObjectNode).getField("sourceId")
      ).toBeInstanceOf(FieldNode);
    });
  });

  describe("on execute object node", () => {
    beforeAll(() => {
      plugin.execute(context.document.getNode("Query") as ObjectNode);
      plugin.execute(context.document.getNode("Todo") as ObjectNode);
    });

    it("creates connection types", () => {
      expect(context.document.getNode("TodoConnection")).toBeInstanceOf(ObjectNode);
      expect(context.document.getNode("TodoEdge")).toBeInstanceOf(ObjectNode);
      expect(context.document.getNode("ResourceConnection")).toBeInstanceOf(ObjectNode);
      expect(context.document.getNode("ResourceEdge")).toBeInstanceOf(ObjectNode);
    });

    it("updates field types", () => {
      const todosField = (context.document.getNode("Query") as ObjectNode).getField("todos");
      const resourcesField = (context.document.getNode("Todo") as ObjectNode).getField("resources");

      expect(todosField?.type.getTypeName()).toBe("TodoConnection");
      expect(todosField?.getArgument("filter")?.type.getTypeName()).toBe("TodoFilterInput");
      expect(todosField?.hasArgument("first")).toBe(true);
      expect(todosField?.hasArgument("after")).toBe(true);
      expect(todosField?.hasArgument("sort")).toBe(true);

      expect(resourcesField?.type.getTypeName()).toBe("ResourceConnection");
      expect(resourcesField?.getArgument("filter")?.type.getTypeName()).toBe(
        "ResourceEdgeFilterInput"
      );
      expect(resourcesField?.hasArgument("first")).toBe(true);
      expect(resourcesField?.hasArgument("after")).toBe(true);
      expect(resourcesField?.hasArgument("sort")).toBe(true);
    });

    it("creates field resolvers", () => {
      expect(context.loader.hasFieldLoader("Todo.resources")).toStrictEqual(true);
    });
  });

  describe("on cleanup node", () => {
    beforeAll(() => {
      plugin.cleanup(context.document.getNode("Query") as ObjectNode);
      plugin.cleanup(context.document.getNode("Todo") as ObjectNode);
    });

    it("removes connection directives from fields", () => {
      const queryNode = context.document.getNode("Query") as ObjectNode;
      const todoNode = context.document.getNode("Todo") as ObjectNode;

      expect(queryNode.getField("me")?.hasDirective("hasOne")).toBeFalsy();
      expect(queryNode.getField("todos")?.hasDirective("hasMany")).toBeFalsy();
      expect(todoNode.getField("resources")?.hasDirective("hasMany")).toBeFalsy();
    });
  });

  describe("on run `after` hook", () => {
    beforeAll(() => {
      plugin.after();
    });

    it("removes connection directive definitions", () => {
      expect(context.document.getNode("ConnectionRelationType")).toBeUndefined();
      expect(context.document.getNode("SortKeyInput")).toBeUndefined();
      expect(context.document.getNode("edges")).toBeUndefined();
      expect(context.document.getNode("node")).toBeUndefined();
    });
  });

  it("matches snapshot", () => {
    expect(context.document.print()).toMatchSnapshot();
  });
});
