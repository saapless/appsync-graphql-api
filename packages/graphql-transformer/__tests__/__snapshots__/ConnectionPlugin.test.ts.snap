// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectionPlugin matches snapshot 1`] = `
"type User {
  id: ID!
  username: String!
}

type Todo {
  id: ID
  content: String
  resources(filter: ResourceEdgeFilterInput, first: Int, after: String, sort: SortDirection): ResourceConnection!
  sourceId: ID @writeOnly
}

type Document {
  url: String!
  sourceId: ID @writeOnly
}

type Message {
  content: String!
  sourceId: ID @writeOnly
}

union Resource = Document | Message

type Query {
  me: User
  todos(filter: TodoFilterInput, first: Int, after: String, sort: SortDirection): TodoConnection!
}

type PageInfo {
  hasNextPage: Boolean
  hasPreviousPage: Boolean
  startCursor: String
  endCursor: String
}

input SizeFilterInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int!]
}

input StringFilterInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  in: [String!]
  contains: String
  notContains: String
  between: [String!]
  beginsWith: String
  attributeExists: Boolean
  size: SizeFilterInput
}

input IntFilterInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  in: [Int!]
  between: [Int!]
  attributeExists: Boolean
}

input FloatFilterInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  in: [Float!]
  between: [Float!]
  attributeExists: Boolean
}

input BooleanFilterInput {
  ne: Boolean
  eq: Boolean
  attributeExists: Boolean
}

input IDFilterInput {
  ne: ID
  eq: ID
  in: [ID!]
  attributeExists: Boolean
}

input ListFilterInput {
  contains: String
  notContains: String
  size: SizeFilterInput
}

enum SortDirection {
  ASC
  DESC
}

type TodoConnection {
  edges: [TodoEdge!]!
  pageInfo: PageInfo!
}

type TodoEdge {
  cursor: String @clientOnly
  node: Todo @clientOnly
}

input TodoFilterInput {
  id: IDFilterInput
  content: StringFilterInput
  and: [TodoFilterInput]
  or: [TodoFilterInput]
  not: TodoFilterInput
}

type ResourceConnection {
  edges: [ResourceEdge!]!
  pageInfo: PageInfo!
  keys: [Node] @serverOnly
}

type ResourceEdge {
  cursor: String @clientOnly
  node: Resource @clientOnly
  id: ID! @serverOnly
  sourceId: ID! @writeOnly
  targetId: ID! @writeOnly
  createdAt: AWSDateTime! @filterOnly
  updatedAt: AWSDateTime! @filterOnly
}

input ResourceEdgeFilterInput {
  createdAt: StringFilterInput
  updatedAt: StringFilterInput
  and: [ResourceEdgeFilterInput]
  or: [ResourceEdgeFilterInput]
  not: ResourceEdgeFilterInput
}

input ResourceEdgeInput {
  sourceId: ID!
  targetId: ID!
}

type Mutation {
  createResourceEdge(input: ResourceEdgeInput!): ResourceEdge
  deleteResourceEdge(input: ResourceEdgeInput!): ResourceEdge
}"
`;
